src/bundle_info.c

```diff
@@ -0,0 +1,570 @@
+
+	string_or_die(&info_header, " Info for bundle: %s ", bundle->component);
+	header_length = strlen(info_header);
+
+	/* info header */
+	print_pattern("_", header_length);
+	print("%s\n", info_header);
+	print_pattern("_", header_length);
+
+	/* status info */
+	print("\nStatus: %s%s\n", bundle_info->is_installed ? "Installed" : "Not installed", bundle_info->is_experimental ? " (experimental)" : "");
+
+	/* version info */
+	if (bundle_info->is_installed) {
+		if (bundle_info->manifest->version < bundle_info->version_latest_available) {
+			print("\nThere is an update for bundle %s:\n", bundle->component);
+		} else {
+			print("Bundle %s is up to date:\n", bundle->component);
+		}
+		print(" - Last updated in version: %d\n", bundle->version);
+		print(" - ");
+	}
+	print("Latest available version: %d\n", bundle_info->version_latest_available);
+
+	/* size info */
+	print("\nBundle size:\n");
+	prettify_size(bundle->contentsize, &pretty_size);
+	print(" - Size of bundle: %s\n", pretty_size);
+	free_string(&pretty_size);
+	if (bundle_info->is_installed) {
+		prettify_size(bundle_info->size_total, &pretty_size);
+		print(" - Size bundle takes on disk (includes dependencies): %s\n", pretty_size);
+		free_string(&pretty_size);
+	} else {
+		if (bundle_info->size_required >= 0) {
+			prettify_size(bundle_info->size_required, &pretty_size);
+			print(" - Size the bundle will take on disk if installed (includes dependencies): %s\n", pretty_size);
+			free_string(&pretty_size);
+		} else {
+			print(" - Size the bundle will take on disk if installed (includes dependencies): unknown\n");
+		}
+	}
+
+	free_string(&info_header);
+}
+
+static void print_includes(struct bundle_info *bundle_info)
+{
+	struct manifest *bundle_manifest = bundle_info->manifest;
+	struct list *iter;
+	char *included_bundle;
+
+	/* print direct includes */
+	if (bundle_manifest->includes || bundle_manifest->optional) {
+		print("\nDirectly included bundles (%d):\n", list_len(bundle_manifest->includes) + list_len(bundle_manifest->optional));
+		iter = list_head(bundle_manifest->includes);
+		if (iter) {
+			while (iter) {
+				included_bundle = iter->data;
+				print(" - %s\n", included_bundle);
+				iter = iter->next;
+			}
+		}
+		iter = list_head(bundle_manifest->optional);
+		if (iter) {
+			while (iter) {
+				included_bundle = iter->data;
+				if (is_installed_bundle(bundle_manifest->component)) {
+					/* it only makes sense to provide details about an optional include
+					 * not being installed if the bundle that includes it is installed */
+					print(" - %s (optional%s)\n", included_bundle, is_installed_bundle(included_bundle) ? "" : ", not installed");
+				} else {
+					print(" - %s (optional)\n", included_bundle);
+				}
+				iter = iter->next;
+			}
+		}
+	}
+
+	/* print indirect includes (if any) */
+	iter = list_head(bundle_info->indirect_includes);
+	if (iter) {
+		print("\nIndirectly included bundles (%d):\n", list_len(bundle_info->indirect_includes));
```

Comment:
yes, we are installing optionals recursively because they are considered in `add_subscriptions` https://github.com/clearlinux/swupd-client/blob/master/src/bundle.c#L782

Reply:
> what if there's indirectly included optionals?

all indirect includes would still be listed but there would be no distinction in if they are optional or not